### ISSUE TITLE
Pointer (on GPU) support

### ISSUE THREAD

**Issue by L3tum** [m4rs-mt/ILGPU#867]
Hey, again, tremendous library. Having much fun with it.

One thing I had an issue with was trees, specifically use of pointers.
In C#, as you probably know, structs cannot self-reference. A solution to that would be to use pointers.

I understand that copying pointers from/to CPU/GPU would be a huge effort to translate and I understand that it's not supported.
However, if a struct only lives on the GPU, I don't see any issue in supporting pointers. Afaik both CUDA and OpenCL support pointers.

So far it seems like ILGPU enters an infinite loop of `GetTypeInfo` when you use a pointer in a struct.

I can provide more info or an example if you want.

My current workaround is to instead save an integer to index into an ArrayView, but that probably leaves a fair amount of performance on the table and also prevents me from modifying the array in a way that would change the indices.

**Comment by MoFtZ** [comment]
Using unsafe pointers should be fine - I have previously used it on the Cuda accelerator.

I was able to reproduce the infinite loop crash. ILGPU does not currently support a self-referencing struct. e.g.
```CSharp
unsafe struct TreeNode
{
    public int Data { get; set; }
    public TreeNode* Left { get; set; }
    public TreeNode* Right { get; set; }
}
```
You could trying tricking ILGPU by internally using `IntPtr` and exposing properties. e.g.
```CSharp
public unsafe struct TreeNode
{
    public int Data { get; set; }
    private IntPtr _Left;
    private IntPtr _Right;

    public TreeNode* Left
    {
        get => (TreeNode*)_Left;
        set => _Left = (IntPtr)value;
    }

    public TreeNode* Right
    {
        get => (TreeNode*)_Right;
        set => _Right = (IntPtr)value;
    }
}
```

Also note that the `ArrayView` struct has a `LoadEffectiveAddressAsPtr` extension method to get the native pointer.

**Comment by L3tum** [comment]
Thank you, I stopped after initially trying the unsafe pointers and didn't think of IntPtr. Would that be a 64-bit (i.e. 8-byte) pointer in that case?

**Comment by MoFtZ** [comment]
The bit-width of `IntPtr` is the same as a pointer. Depends on the bitness of the process.

**Comment by L3tum** [comment]
Coolio, thanks @MoFtZ ! One last question: Is there a way to dump the code for a specialized kernel? I don't want to rip the specialization out but so far haven't found a way to dump it either

**Comment by L3tum** [comment]
Oh, also the `LoadEffectiveAddressAsPtr`  is not available in a Kernel according the the remark and attribute `This method is not supported on accelerators.`.

Is there a way to get a pointer to an object in a Kernel? I thought that'd be easier and I'm not sure if moving an IntPtr from CPU to GPU is supported by ILGPU?

**Comment by L3tum** [comment]
Hmm, your proposed code doesn't seem to pass through ILGPU. I'm not entirely sure what the problem is considering the functions individually compile fine, but put together crashes the compiler. Because the exception is rethrown I also can't step through the Stacktrace that well. Here's the message at least: `The given key 'const_1752: True [Int1]' was not present in the dictionary.` To be clear I'm only using `bool` as a function return type, which had worked fine so far.

It seems like certain combinations of methods crash the compiler with various exceptions. One I just got as well is this: `Index was outside the bounds of the array.` which seems to be related cause it was thrown here: 
````
   at System.Collections.Immutable.ImmutableArray`1.get_Item(Int32 index)
   at ILGPU.IR.Method.ParameterCollection.get_Item(Int32 index)
   at ILGPU.IR.Transformations.InferKernelAddressSpaces.Rewrite(RewriterContext context, MethodDataProvider data, MethodCall call)
   at ILGPU.IR.Rewriting.Rewriter`4.<>c__DisplayClass11_0`1.<Add>b__1(TContext& context, T data, Value value)
   at ILGPU.IR.Rewriting.Rewriter`1.Rewrite(BasicBlockCollection`2& blocks, Builder builder, T data)
   at ILGPU.IR.Transformations.InferKernelAddressSpaces.PerformTransformation(IRContext context, Builder builder, MethodDataProvider& intermediate, Landscape landscape, Entry current)
   at ILGPU.IR.Transformations.OrderedTransformation`1.Transform(MethodCollection& methods)
   at ILGPU.IR.IRContext.Transform(Transformer& transformer)
   at ILGPU.Backends.Backend.Compile[TBackendHook](Method kernelMethod, EntryPointDescription& entry, KernelSpecialization& specialization, TBackendHook backendHook)
````

To be clear, the objects are constructed in GPU buffer and the addresses are taken in a Kernel and used in a Kernel.

**Comment by L3tum** [comment]
Okay so debugging a bit more it seems like the following line is the culprit:
````c#
if (HasBoundingVolume == 0 && Atomic.CompareExchange(ref HasBoundingVolume, 0, 1) == 0)
````

My thoughts were to first do it non-atomically to compare since the only actual relevant atomic operation is the exchange, not necessarily the comparison. But that breaks the compiler apparently.

Having solved that some other method now creates an issue, although I'm 99% certain that it only uncovers a deeper issue.

The exception:

````
Index was outside the bounds of the array.
   at ILGPU.IR.BasicBlockMap`1.get_Item(BasicBlock block)
   at ILGPU.Backends.PhiBindings.Create[TOrder,TDirection,TAllocator](BasicBlockCollection`2& collection, TAllocator allocator)
   at ILGPU.Backends.OpenCL.CLCodeGenerator.GenerateCodeInternal()
   at ILGPU.Backends.OpenCL.CLKernelFunctionGenerator.GenerateCode()
   at ILGPU.Backends.CodeGeneratorBackend`4.<>c__DisplayClass1_0.<Compile>b__0(Int32 i)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
````

With the code in question triggering it being:

````c#
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public bool Hit(float tMin, float tMax, ref Ray ray)
    {
        float t0;
        float t1;

        t0 = Minimum.X - ray.Origin.X * ray.InverseDirection.X;
        t1 = Maximum.X - ray.Origin.X * ray.InverseDirection.X;
        tMin = XMath.Min(XMath.Max(t0, tMin), XMath.Max(t1, tMin));
        tMax = XMath.Max(XMath.Min(t0, tMax), XMath.Min(t1, tMax));
        if (tMax <= tMin)
        {
            return false;
        }

        t0 = Minimum.Y - ray.Origin.Y * ray.InverseDirection.Y;
        t1 = Maximum.Y - ray.Origin.Y * ray.InverseDirection.Y;
        tMin = XMath.Min(XMath.Max(t0, tMin), XMath.Max(t1, tMin));
        tMax = XMath.Max(XMath.Min(t0, tMax), XMath.Min(t1, tMax));
        if (tMax <= tMin)
        {
            return false;
        }

        t0 = Minimum.Z - ray.Origin.Z * ray.InverseDirection.Z;
        t1 = Maximum.Z - ray.Origin.Z * ray.InverseDirection.Z;
        tMin = XMath.Min(XMath.Max(t0, tMin), XMath.Max(t1, tMin));
        tMax = XMath.Max(XMath.Min(t0, tMax), XMath.Min(t1, tMax));
        if (tMax <= tMin)
        {
            return false;
        }

        // for (var a = 0; a < 3; a++)
        // {
        //     var t0 = (At(Minimum, a) - At(ray.Origin, a)) * At(ray.InverseDirection, a);
        //     var t1 = (At(Maximum, a) - At(ray.Origin, a)) * At(ray.InverseDirection, a);
        //     tMin = XMath.Min(XMath.Max(t0, tMin), XMath.Max(t1, tMin));
        //     tMax = XMath.Max(XMath.Min(t0, tMax), XMath.Min(t1, tMax));
        //     if (tMax <= tMin)
        //     {
        //         return false;
        //     }
        // }

        return true;
    }
````

And the code making use of it

````c#
    public bool Hit(float tMin, ref Ray ray, ref HitRecord hitRecord)
    {
        var hasHit = false;

        if (BoundingVolume.Hit(tMin, hitRecord.Time, ref ray))
        {
            unsafe
            {
                if (LeftType == NodeType.SPHERE)
                {
                    if (LeftHittable->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
                else
                {
                    if (LeftNode->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
            
                if (RightType == NodeType.SPHERE)
                {
                    if (RightHittable->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
                else
                {
                    if (RightNode->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
            }
        }

        return hasHit;
    }
````

If I change that code to this, then it compiles fine:

````c#
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public bool Hit(float tMin, ref Ray ray, ref HitRecord hitRecord)
    {
        var hasHit = false;

        hasHit = BoundingVolume.Hit(tMin, hitRecord.Time, ref ray);

        if (false)
        {
            unsafe
            {
                if (LeftType == NodeType.SPHERE)
                {
                    if (LeftHittable->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
                else
                {
                    if (LeftNode->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
            
                if (RightType == NodeType.SPHERE)
                {
                    if (RightHittable->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
                else
                {
                    if (RightNode->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }
            }
        }

        return hasHit;
    }
````

**Comment by L3tum** [comment]
Seems like the cast is the problem. Reinterpreting the same pointer into different objects. This code also doesn't work for example, but only using the Sphere Code does.

````c#
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public bool Hit(float tMin, ref Ray ray, ref HitRecord hitRecord)
    {
        unsafe
        {
            if (BoundingVolume.Hit(tMin, hitRecord.Time, ref ray))
            {
                bool hasHit = false;
                
                if (LeftType == NodeType.SPHERE)
                {
                    if (LeftHittable->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }

                if (RightType == NodeType.BVHNODE)
                {
                    if (RightNode->Hit(tMin, ref ray, ref hitRecord))
                    {
                        hasHit = true;
                    }
                }

                return hasHit;
            }
        }
        
        return false;
    }
````

**Comment by L3tum** [comment]
I got it to compile but it got into an infinite loop at point. Executing the same code on the CPU is absolutely no problem so I'd guess there's some miscompilation there, or perhaps some issues with the pointers. I'd love to be able to provide more information but short of dumping my whole project on you I can't really do much. Haven't even managed to dump the OpenCL code

**Comment by L3tum** [comment]
It seems like whatever the program does now on the GPU crashes my driver hard, and prevents another kernel from launching again. Stopping and restarting the program makes it hang for a second before throwing an Exception while trying to create an Accelerator. 

My best guess would be that the memory address is wrong on the GPU and it gets somewhere it shouldn't, because I otherwise didn't change that much substantial code between the pointer version and the indexed version.

**Comment by L3tum** [comment]
So I'm trying to dump the kernel with some debugging info now... It's now throwing the following exception, which, to me, seems to be a contradiction...

`System.NotSupportedException: The allocation size of type 'Struct<Float32 [0], Float32 [4], Float32 [8], Float32 [12], Float32 [16], Float32 [20], Int32 [24], Int32 [28], Int32 [32], Int8 [36], Int8 [37]>' must be statically known and not a dynamic value`.

**Comment by L3tum** [comment]
Whatever the compiler is doing it seems to completely wreck my system and basically at random. I can run the code fine 5 times, but the sixth one will crash my system without any changes. I can upload the project I guess for you to look at but I doubt it'd help much. I managed to dump the OpenCL code and it's 4000 lines long (the source C# code is <600 lines long) and seems to contain various miscompilations, such as allocating 6 `long`s and then only using one of them...while my code doesn't contain any `long` fields at all. I'm sure the optimimzer would remove these statements but I'm not sure where they're coming from in the first place. That's all without any pointer usage mind you, that's the base indexed version that seems to work fine most of the time but sometimes doesn't. 

Most of the issues seem to stem from some kind of inlining, as far as I can tell. Depending on what I use, for example a `switch` statement with more than 3 arms, there'll be other issues or even exceptions thrown during compilations. Similarly deactivating the inlining of other methods also seems to fix certain problems almost at random, as if there's some kind of buffer overflow or such from having too many methods inlined. Completely disabling the inlining makes the compiler throw exceptions as well though (see the comment above) at places that make no sense to my (admittedly, not very trained) eye. 

I've written two different applications, one CLI dumping the final result into an image, the other displaying the image in a WPF window. The first one crashes the system while the second one works fine, with absolutely no difference to the actual code or usage thereof. The only difference would be accessing the memory buffer once per iteration in the WPF version, and once per execution in the CLI version, but these things are properly synchronized and even locked with a Mutex so there wouldn't be any simultaneous access. Aside from that the WPF version is the one working fine, where I'd expect the most problems otherwise. 

I'm sorry, I really like what you're doing with this library, but at this point I've spent well over 3 days chasing weird bugs happening seemingly at random and with no discernible difference of the code at hand, across multiple versions and even nightly versions. Maybe I'll check back in a year or so but for now I'm definitely done.
